active/003-seq-feature.md

```diff
@@ -0,0 +1,140 @@
+
+# Motivation
+
+The current implementation of `positional_metadata` in the `Sequence` class takes > 1TB memory to read in a E. coli genome with 260K features. And the `pandas` sparse data frame can alleviate the mem usage, but it is still too memory hogging. The discussion is mainly in this [thread](https://github.com/biocore/scikit-bio/issues/1159).  Storing intervals instead of `positional_metadata` would substantially save on memory.
+
+Additional benefits of having a `feature_metadata` objects include the ability to rapidly query features by coordinates.  Interval trees allow for both querying by position and ranges to retrieve all overlapping intervals.  This can be particularly helpful when dealing with layers of features (i.e. transcripts made up of exons, or operons made up of genes).
+
+# Detailed design
+
+We propose 2 new public data objects: `BoundFeature` and `IntervalMetadata`. `BoundFeature` stores all the attributes of a feature. `IntervalMetadata` stores all the `BoundFeatures` objects of a sequence and add it as `feature_metadata` attribute (as similar to `positional_metadata`) in `Sequence` (and its child classes).
+
+## `BoundFeature` object
+This object is a *mutable* object, that contains arbitrary attributes (i.e. `gene_name`, `product`, ...) to store the all the info of a sequence feature. This object would also have a [weakref](https://docs.python.org/3/library/weakref.html) to the corresponding `IntervalMetadata` object. If the intervals of a `BoundFeature` are updated, the interval tree within the `IntervalMetadata` object is also auto updated. The weak reference enables the tight coupling between `BoundFeature` and `IntervalMetadata`. The mutability of `BoundFeature` enables us to directly modify a `BoundFeature` object from a `IntervalMetadata` object.
+
+### enforced attributes
+Besides user arbitrarily given attributes, we enforce the following attribues:
+
+* `intervals`: store intervals of coordinates.  This is represented as a list of tuples of a pair of ints
+* `strand`: strand
+* `wref`: a weak reference to `IntervalMetadata` object.  This is a private attribute.
+* `boundaries`: records the openness of each interval.  So a boundary of `(True, False)` would it indicate that the exact right boundary is unknown, corresponding to the examples [here](ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3).
+
+
+### methods
+`__init__(**kwargs)`
+The construction would be like:
+```python
+>>> f = BoundFeature(intervals=[1, (4, 7)], strand='+', boundaries=[(True, True), (False, False)], wref=None, gene='sagA', function='toxin')
+>>> f.intervals  # get coordinates
+>>> f.wref   # get weak ref to the interval metadata.  This is a private attribute
+>>> f.function   # get the feature info
+```
+Note, in the above example, the interval `1` is shorthand for `(1, 2)`.
```

Comment:
Drop it?

On Wed, May 4, 2016 at 9:42 AM, Daniel McDonald <notifications@github.com>
wrote:

> In active/003-seq-feature.md
> <https://github.com/biocore/scikit-bio-rfcs/pull/9#discussion_r62072714>:
>
> > +* `intervals`: store intervals of coordinates.  This is represented as a list of tuples of a pair of ints
> > +* `strand`: strand
> > +* `wref`: a weak reference to `IntervalMetadata` object.  This is a private attribute.
> > +* `boundaries`: records the openness of each interval.  So a boundary of `(True, False)` would it indicate that the exact right boundary is unknown, corresponding to the examples [here](ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3).
> > +
> > +
> > +### methods
> > +`__init__(**kwargs)`
> > +The construction would be like:
> > +```python
> > +>>> f = BoundFeature(intervals=[1, (4, 7)], strand='+', boundaries=[(True, True), (False, False)], wref=None, gene='sagA', function='toxin')
> > +>>> f.intervals  # get coordinates
> > +>>> f.wref   # get weak ref to the interval metadata.  This is a private attribute
> > +>>> f.function   # get the feature info
> > +```
> > +Note, in the above example, the interval `1` is shorthand for `(1, 2)`.
>
> This comes at a cost of semantic confusion though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio-rfcs/pull/9/files/9a8ee7cb3113c5aeb7bf64f23b38db76859e0db7#r62072714>
>